apps/emqx/src/emqx_trace/emqx_trace.erl

```diff
@@ -261,22 +304,136 @@ clear() ->
+        {ok, Trace = #?TRACE{start_at = Start}} ->
+            _ = emqx_trace_handler:filesync(mk_handler_id(Trace)),
+            Basename = log_filepath(Name, Start),
+            case emqx_trace_handler:find_log_fragment(first, Basename) of
+                {ok, _Filename, _Info, Fragment} ->
+                    %% NOTE: Enough information to _safely_ reconstruct the basename.
+                    Cursor = #{s => Start, p => 0, f => Fragment},
+                    stream_log_cursor(Name, Cursor, Limit);
+                none ->
+                    {error, {file_error, enoent}};
+                {error, Reason} ->
+                    {error, {file_error, Reason}}
+            end;
+        Error ->
+            Error
+    end;
+stream_log(Name, {cont, Cursor}, Limit) when Limit > 0 ->
+    stream_log_cursor(Name, Cursor, Limit).
+
+stream_log_cursor(Name, Cursor0 = #{s := Start, p := Pos, f := Fragment0}, Limit) ->
+    NBytes = emqx_maybe:define(Limit, ?DEFAULT_STREAM_CHUNK_SIZE),
+    Basename = log_filepath(Name, Start),
+    case emqx_trace_handler:read_log_fragment_at(Fragment0, Basename, Pos, NBytes) of
+        {ok, Chunk, Fragment} ->
+            Size = byte_size(Chunk),
+            Cursor = Cursor0#{p := Pos + Size, f := Fragment},
+            case Size of
+                N when N >= NBytes ->
+                    Cont = {cont, Cursor};
+                _Partial ->
+                    case emqx_trace_handler:find_log_fragment({next, Fragment}, Basename) of
+                        {ok, _, _, FragmentNext} ->
+                            Cont = {cont, Cursor#{p := 0, f := FragmentNext}};
+                        none ->
+                            Cont = {eof, Cursor};
+                        {error, enoent} ->
+                            %% Here it can actually mean active log rotation, ask the caller to
+                            %% retry. If the error persists, fragments were probably manually
+                            %% deleted.
+                            %% See `emqx_trace_handler:find_log_fragment/2`.
+                            Cont = {retry, Cursor};
+                        {error, Reason} ->
+                            Cont = {cont, #{e => {file_error, Reason}}}
+                    end
+            end,
+            {ok, Chunk, Cont};
+        {error, enoent} ->
+            %% Same here, see `emqx_trace_handler:find_log_fragment/2`.
+            {ok, <<>>, {retry, Cursor0}};
+        {error, Reason} ->
+            {error, {file_error, Reason}}
+    end;
+stream_log_cursor(_Name, #{e := Reason}, _Limit) ->
+    {error, Reason};
+stream_log_cursor(_Name, _Cursor, _Limit) ->
+    {error, bad_cursor}.
+
+log_filepath(Name, Start) ->
+    filename:join(trace_dir(), log_filename(Name, Start)).
+
+log_filename(Name, Start) ->
+    [Time, _] = string:split(calendar:system_time_to_rfc3339(Start), "T", leading),
+    lists:flatten(["trace_", binary_to_list(Name), "_", Time, ".log"]).
```

Comment:
Q: is there any constraint on the name?  If it can be any string, maybe we should call `emqx_utils:safe_filename` on the result.

Reply:
Yes, and it's actually both [here](https://github.com/emqx/emqx/blob/5d731e884a28374eb335ceb8e535e8484e918fb3/apps/emqx/src/emqx_trace/emqx_trace.erl#L763-L764) and in the API handler for some weird reason.